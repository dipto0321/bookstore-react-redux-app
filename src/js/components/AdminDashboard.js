import React from 'react';
import {
  Tab, Tabs, TabList, TabPanel,
} from 'react-tabs';
import 'react-tabs/style/react-tabs.scss';
import SearchForm from './SearchForm';

const AdminDashboard = () => (
  <Tabs>
    <TabList>
      <Tab>Manage Users</Tab>
      <Tab>Manage Books</Tab>
    </TabList>
    <TabPanel>
      <h2>Users</h2>
    </TabPanel>
    <TabPanel>
      <SearchForm />
    </TabPanel>
  </Tabs>
);

export default AdminDashboard;
