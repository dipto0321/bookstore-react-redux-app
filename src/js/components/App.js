import React from 'react';
import { Route, Router, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createBrowserHistory } from 'history';
import decode from 'jwt-decode';
import BooksDashboard from './BooksDashboard';
import '../../scss/main.scss';
import LoginPage from '../containers/LoginPage';
import configureStore from '../store/configureStore';
import { setCurrentUser } from '../actions/currentUser';
import { setAuthorizationToken } from '../services/api';
import withAuth from '../hocs/withAuth';

export const history = createBrowserHistory();

const store = configureStore();

if (localStorage.token) {
  const userData = decode(localStorage.token);
  setAuthorizationToken(localStorage.token);

  try {
    store.dispatch(setCurrentUser({ authenticated: true, data: userData }));
  } catch (err) {
    store.dispatch(setCurrentUser({ authenticated: false, data: null }));
  }
}

const App = () => (
  <Provider store={store}>
    <Router history={history}>
      <Switch>
        <Route path="/login" component={LoginPage} />
        <Route path="/" component={withAuth(BooksDashboard, store.getState().currentUser.authenticated)} />
      </Switch>
    </Router>
  </Provider>
);

export default App;
