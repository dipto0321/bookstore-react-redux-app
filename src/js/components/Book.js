import React from 'react';
import PropTypes from 'prop-types';

const Book = (props) => {
  const { handleRemove, book } = props;
  const { id, title, category } = book;

  return (
    <div className="card">
      <div className="card-body">
        <div className="columns">
          <div className="col">
            <h4 className="genre">{category}</h4>
            <h3 className="title">{title}</h3>
            <div className="action">
              <button type="button" onClick={() => handleRemove(id)}>
                Remove
              </button>
            </div>
          </div>
          <div className="col" />
        </div>
      </div>
    </div>
  );
};

Book.propTypes = {
  book: PropTypes.instanceOf(Object).isRequired,
  handleRemove: PropTypes.func.isRequired,
};

export default Book;
