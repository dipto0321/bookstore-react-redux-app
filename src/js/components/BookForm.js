import React, { useState } from 'react';
import { connect } from 'react-redux';
import { addBookToLibrary } from '../thunks/book';
import { switchAddBookMode } from '../actions/book';
import InputWrapper from './InputWrapper';

export const BookForm = ({ book, addBookToLibrary, switchAddBookMode }) => {
  const [title, setTitle] = useState(book ? book.title : '');
  const [subTitle, setSubTitle] = useState(book ? book.subtitle : '');
  const [authors, setAuthors] = useState(book ? book.authors.join(', ') : '');
  const [description, setDescription] = useState(book ? book.description : '');
  const [publishedDate, setPublishedDate] = useState(book ? book.publishedDate : '');
  const [thumbNail, setThumbNail] = useState(book ? book.imageLinks.smallThumbnail : '');
  const [isbn, setIsbn] = useState(book ? book.industryIdentifiers[0].identifier : '');
  const [pageCount, setPageCount] = useState(book ? book.pageCount : '');
  const [category, setCategory] = useState(book ? book.categories[0] : '');

  const handleSubmit = (e) => {
    e.preventDefault();
    addBookToLibrary({
      book: {
        title,
        subtitle: subTitle,
        authors,
        description,
        published_date: publishedDate,
        thumbnail: thumbNail,
        isbn,
        page_count: pageCount,
        category,
      },
    })
      .then(() => switchAddBookMode());
  };

  return (
    <form>
      <InputWrapper
        inputValue={title}
        labelValue="Title: "
        setInput={setTitle}
        type="text"
        inputId="title"
        error={null}
      />
      <InputWrapper
        inputValue={subTitle}
        labelValue="Sub-title: "
        setInput={setSubTitle}
        type="text"
        inputId="sub-title"
        error={null}
      />
      <InputWrapper
        inputValue={authors}
        labelValue="Authors: "
        setInput={setAuthors}
        type="text"
        inputId="authors"
        error={null}
      />
      <InputWrapper
        inputValue={category}
        labelValue="Category: "
        setInput={setCategory}
        type="text"
        inputId="category"
        error={null}
      />
      <InputWrapper
        inputValue={isbn}
        labelValue="ISBN: "
        setInput={setIsbn}
        type="text"
        inputId="isbn"
        error={null}
      />
      <InputWrapper
        inputValue={description}
        labelValue="Description: "
        setInput={setDescription}
        type="textarea"
        inputId="description"
        error={null}
      />
      <InputWrapper
        inputValue={publishedDate}
        labelValue="Published Date: "
        setInput={setPublishedDate}
        type="text"
        inputId="published-date"
        error={null}
      />
      <InputWrapper
        inputValue={pageCount}
        labelValue="Page Count: "
        setInput={setPageCount}
        type="text"
        inputId="page-count"
        error={null}
      />
      <InputWrapper
        inputValue={thumbNail}
        labelValue="Thumbnail Link: "
        setInput={setThumbNail}
        type="text"
        inputId="thumbnail"
        error={null}
      />
      <div className="btn-wrapper">
        <button
          type="submit"
          className="add-book-btn"
          onClick={handleSubmit}
        >
          Add Book To Library
        </button>
      </div>
    </form>
  );
};

const mapStateToProps = state => ({
  book: state.addBookMode.book,
});

export default connect(mapStateToProps, { addBookToLibrary, switchAddBookMode })(BookForm);
