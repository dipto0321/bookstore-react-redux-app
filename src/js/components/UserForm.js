import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { signUp, updateAccount } from '../thunks/user';
import history from '../services/history';
import { setErrors } from '../actions/errors';
import InputWrapper from './InputWrapper';

export const UserForm = ({
  signUp,
  errors,
  setErrors,
  isAuthenticated,
  userData,
  updateAccount,
}) => {
  const [username, setUsername] = useState(userData ? userData.username : '');
  const [email, setEmail] = useState(userData ? userData.email : '');
  const [password, setPassword] = useState('');
  const [passwordConfirmation, setPasswordConfirmation] = useState('');
  const [firstName, setFirstName] = useState(userData ? userData.first_name : '');
  const [lastName, setLastName] = useState(userData ? userData.last_name : '');
  const saveUser = isAuthenticated ? updateAccount : signUp;

  const removeErrorMsg = useCallback(() => {
    setErrors(null);
  }, [setErrors]);

  useEffect(() => {
    removeErrorMsg();
    return () => setErrors(null);
  }, [removeErrorMsg, setErrors]);

  const handleSubmit = (e) => {
    e.preventDefault();
    const secureAttributes = { password, password_confirmation: passwordConfirmation };
    const profileAttributes = {
      username, email, first_name: firstName, last_name: lastName,
    };
    const userParamsData = password === '' ? profileAttributes : { ...profileAttributes, ...secureAttributes };
    const params = userData ? { user: userParamsData, id: userData.id } : { user: userParamsData };
    const usernameParam = isAuthenticated ? userData.username : '';
    saveUser(params, usernameParam)
      .then(() => history.push('/'))
      .catch(err => err);
  };

  return (
    <div>
      <form>
        {
            errors && <div className="error">{errors.message}</div>
          }
        <InputWrapper
          inputValue={username}
          labelValue="Username: "
          setInput={setUsername}
          type="text"
          inputId="username"
          error={errors ? errors.errors.username : null}

        />
        <InputWrapper
          inputValue={email}
          labelValue="Email: "
          setInput={setEmail}
          type="email"
          inputId="email"
          error={errors ? errors.errors.email : null}
        />
        <InputWrapper
          inputValue={password}
          labelValue="Password: "
          setInput={setPassword}
          type="password"
          inputId="password"
          error={errors ? errors.errors.password : null}

        />
        <InputWrapper
          inputValue={passwordConfirmation}
          labelValue="Confirm Password: "
          setInput={setPasswordConfirmation}
          type="password"
          inputId="password-confirmation"
          error={errors ? errors.errors.password : null}

        />
        <InputWrapper
          inputValue={firstName}
          labelValue="First Name: "
          setInput={setFirstName}
          type="text"
          inputId="first-name"
          error={errors ? errors.errors.first_name : null}

        />
        <InputWrapper
          inputValue={lastName}
          labelValue="Last Name: "
          setInput={setLastName}
          type="text"
          inputId="last-name"
          error={errors ? errors.errors.last_name : null}

        />
        <div className="btn-wrapper">
          <button
            type="submit"
            className="user-form__btn"
            onClick={handleSubmit}
          >
            { isAuthenticated ? 'Update Account' : 'Create Account'}
          </button>
        </div>
      </form>
    </div>
  );
};

const mapStateToProps = state => ({
  errors: state.errors,
  isAuthenticated: state.currentUser.authenticated,
  userData: state.currentUser.data,
});

UserForm.propTypes = {
  signUp: PropTypes.func.isRequired,
  errors: PropTypes.instanceOf(Object),
  setErrors: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  userData: PropTypes.instanceOf(Object),
  updateAccount: PropTypes.func.isRequired,
};

UserForm.defaultProps = {
  errors: null,
  userData: null,
};

export default connect(mapStateToProps, { signUp, setErrors, updateAccount })(UserForm);
