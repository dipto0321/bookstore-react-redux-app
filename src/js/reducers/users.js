import { LIST_USERS } from '../actions/actionTypes';

export default (state = [], action) => {
  if (action.type === LIST_USERS) {
    return action.users;
  }
  return state;
};
