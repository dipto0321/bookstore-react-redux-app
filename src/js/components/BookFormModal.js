import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Modal from 'react-modal';
import BookForm from './BookForm';
import { switchAddBookMode } from '../actions/book';

export const BookFormModal = ({ switchAddBookMode, addBookMode }) => (
  <Modal
    isOpen={addBookMode.on}
    ariaHideApp={false}
    onRequestClose={() => switchAddBookMode(addBookMode.book)}
    closeTimeoutMS={200}
  >
    <BookForm />
  </Modal>
);

const mapStateToProps = state => ({
  addBookMode: state.addBookMode,
});

export default connect(mapStateToProps, { switchAddBookMode })(BookFormModal);
