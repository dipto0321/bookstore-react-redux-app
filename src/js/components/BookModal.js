import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import BookForm from './BookForm';
import Book from './Book';
import { switchAddBookMode } from '../actions/book';

export const BookModal = ({ switchAddBookMode, addBookMode, isAdmin }) => (
  <Modal
    isOpen={addBookMode.on}
    ariaHideApp={false}
    onRequestClose={() => switchAddBookMode(addBookMode.book)}
    closeTimeoutMS={200}
  >
    {isAdmin && <BookForm /> }
    {!isAdmin && addBookMode.on && <Book book={addBookMode.book} /> }
  </Modal>
);

const mapStateToProps = state => ({
  addBookMode: state.addBookMode,
});

export default connect(mapStateToProps, { switchAddBookMode })(BookModal);
