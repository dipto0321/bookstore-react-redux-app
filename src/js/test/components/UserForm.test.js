import React from 'react';
import { shallow } from 'enzyme';
import { UserForm } from '../../components/UserForm';
import users from '../fixtures/users';

let wrapper;
const signUp = jest.fn();
const updateAccount = jest.fn();
const setError = jest.fn();

describe('UserForm component for signup', () => {
  beforeEach(() => {
    wrapper = shallow(
      <UserForm
        error={null}
        signUp={signUp}
        setError={setError}
      />,
    );
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call signUp on complete user information', () => {
    const {
      username, email, first_name, last_name,
    } = users[0];

    wrapper.find('InputWrapper').at(0).prop('setInput')(username);
    wrapper.find('InputWrapper').at(1).prop('setInput')(email);
    wrapper.find('InputWrapper').at(2).prop('setInput')('password');
    wrapper.find('InputWrapper').at(3).prop('setInput')('password');
    wrapper.find('InputWrapper').at(4).prop('setInput')(first_name);
    wrapper.find('InputWrapper').at(5).prop('setInput')(last_name);
    wrapper.find('button[type="submit"]').simulate('click', { preventDefault: () => {} });

    expect(signUp).toHaveBeenLastCalledWith({
      user: {
        username,
        email,
        first_name,
        last_name,
        password: 'password',
        password_confirmation: 'password',
      },
    });
  });

  test('should show error when user information incomplete', () => {
    wrapper.setProps({ errors: { message: 'Cannot save user', errors: { username: ['already taken'] } } });
    expect(wrapper.find('div[className="error"]').exists()).toBeTruthy();
  });
});

describe('UserForm for editing user account', () => {
  const {
    id, username, email, first_name, last_name,
  } = users[0];
  beforeEach(() => {
    wrapper = shallow(
      <UserForm
        error={null}
        updateAccount={updateAccount}
        setError={setError}
        isAuthenticated
        userData={{
          id, username, email, first_name, last_name,
        }}
      />,
    );
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call updateAccount on successful account update', () => {
    const newFirstName = 'Rocky';
    wrapper.find('InputWrapper').at(4).prop('setInput')(newFirstName);
    wrapper.find('button[type="submit"]').simulate('click', { preventDefault: () => {} });
    expect(updateAccount).toHaveBeenLastCalledWith({
      user: {
        username, email, first_name: newFirstName, last_name,
      },
    });
  });

  test('should show an error when a user info is missing', () => {
    wrapper.setProps({ errors: { message: 'Cannot update user', errors: { email: ["can't be blank"] } } });
    wrapper.find('button[type="submit"]').simulate('click', { preventDefault: () => {} });
    expect(wrapper.find('div[className="error"]').exists()).toBeTruthy();
  });
});
