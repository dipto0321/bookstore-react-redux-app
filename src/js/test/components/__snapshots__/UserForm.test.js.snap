// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm component for signup should render correctly 1`] = `
<div>
  <form>
    <InputWrapper
      inputId="username"
      inputValue=""
      labelValue="Username: "
      setInput={[Function]}
      type="text"
    />
    <InputWrapper
      inputId="email"
      inputValue=""
      labelValue="Email: "
      setInput={[Function]}
      type="email"
    />
    <InputWrapper
      inputId="password"
      inputValue=""
      labelValue="Password: "
      setInput={[Function]}
      type="password"
    />
    <InputWrapper
      inputId="password-confirmation"
      inputValue=""
      labelValue="Confirm Password: "
      setInput={[Function]}
      type="password"
    />
    <InputWrapper
      inputId="first-name"
      inputValue=""
      labelValue="First Name: "
      setInput={[Function]}
      type="text"
    />
    <InputWrapper
      inputId="last-name"
      inputValue=""
      labelValue="Last Name: "
      setInput={[Function]}
      type="text"
    />
    <div
      className="btn-wrapper"
    >
      <button
        className="user-form__btn"
        onClick={[Function]}
        type="submit"
      >
        Create Account
      </button>
    </div>
  </form>
</div>
`;

exports[`UserForm for editing user account should render correctly 1`] = `
<div>
  <form>
    <InputWrapper
      inputId="username"
      inputValue="ryto"
      labelValue="Username: "
      setInput={[Function]}
      type="text"
    />
    <InputWrapper
      inputId="email"
      inputValue="ryto@gmail.com"
      labelValue="Email: "
      setInput={[Function]}
      type="email"
    />
    <InputWrapper
      inputId="password"
      inputValue=""
      labelValue="Password: "
      setInput={[Function]}
      type="password"
    />
    <InputWrapper
      inputId="password-confirmation"
      inputValue=""
      labelValue="Confirm Password: "
      setInput={[Function]}
      type="password"
    />
    <InputWrapper
      inputId="first-name"
      inputValue="Ryto"
      labelValue="First Name: "
      setInput={[Function]}
      type="text"
    />
    <InputWrapper
      inputId="last-name"
      inputValue="Verkar"
      labelValue="Last Name: "
      setInput={[Function]}
      type="text"
    />
    <div
      className="btn-wrapper"
    >
      <button
        className="user-form__btn"
        onClick={[Function]}
        type="submit"
      >
        Update Account
      </button>
    </div>
  </form>
</div>
`;
